### PR TITLE
[Graphbolt] Add pickle methods for CSCSamplingGraph

### PR DESCRIPTION
**Pull request by RamonZhou**
## Description
Implemented `__getstate__` and `__setstate__` methods for CSCSamplingGraph.
Solve issue #6169.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- Add `GetState` and `SetState`.
- Bind torch scripts.
- Add tests.

### CODE REVIEW

File: graphbolt/src/csc_sampling_graph.cc [hunk 1]
```diff
@@ -95,6 +95,39 @@ void CSCSamplingGraph::Save(torch::serialize::OutputArchive& archive) const {
   }
 }
 
+void CSCSamplingGraph::SetState(
+    std::vector<torch::Dict<std::string, torch::Tensor>>& state) {
+  TORCH_CHECK(state.size() >= 1 && state.size() <= 2, "Wrong length of state.");
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:100]: why the magic number is 1 or 2? Maybe add a comment?

File: tests/python/pytorch/graphbolt/test_multi_process_dataloader.py [hunk 2]
```diff
@@ -27,7 +27,6 @@ def fetch_func(features, labels, data):
     return input_features, output_labels, adjs
 
 
-@unittest.skipIf(os.name == "nt", reason="Do not support windows yet")
 # TODO (peizhou): Will enable windows test once CSCSamplingraph is pickleable.
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_multi_process_dataloader.py:30]: Also remove the TODO.

File: graphbolt/src/csc_sampling_graph.cc (hunk 1 above)
Comment (RamonZhou) [reply graphbolt/src/csc_sampling_graph.cc:102]: It's not magic number but the length of the vector. I've added some comments.

**Comment by frozenbugs** [comment]
Overall LGTM, @czkkkkkk please give approve if you are good with this PR.

File: tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py [hunk 3]
```diff
@@ -252,6 +254,77 @@ def test_load_save_hetero_graph(num_nodes, num_edges, num_ntypes, num_etypes):
     assert graph.metadata.edge_type_to_id == graph2.metadata.edge_type_to_id
 
 
+@unittest.skipIf(
```
Comment (czkkkkkk) [inline tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:257]: Could you test whether the pickle function work well with pytorch multiprocess spawn? You may use the CSCSamplingGraph as the argument of the pickle function.

File: graphbolt/src/csc_sampling_graph.cc (hunk 1 above)
Comment (czkkkkkk) [reply graphbolt/src/csc_sampling_graph.cc:102]: It depends on how we define the serialized data structure. I think it is better to use `torch::Dict<std::string, torch::Dict<std::string, torch::Tensor>>`, where the keys of the outer `Dict` are something like `regular_tensor` and `edge_attribute`.

File: graphbolt/src/csc_sampling_graph.cc [hunk 4]
```diff
@@ -95,6 +95,48 @@ void CSCSamplingGraph::Save(torch::serialize::OutputArchive& archive) const {
...
+  if (state.find("edge_attributes") != state.end()) {
+    edge_attributes_ = state.at("edge_attributes");
+  }
+}
+
+torch::Dict<std::string, torch::Dict<std::string, torch::Tensor>>
+CSCSamplingGraph::GetState() const {
+  // State is a vector of dicts. The first dict contains all the tensor
+  // attributes. If edge_attributes is not None, it will be stored as the second
+  // dict.
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:121]: what will happen in the future if we add another dict of tensor in CSCSamplingGraph?

Comment (frozenbugs) [reply graphbolt/src/csc_sampling_graph.cc:121]: It seems that the comment if out of data? The state is a dict of dict in your code.

Comment (RamonZhou) [reply graphbolt/src/csc_sampling_graph.cc]: The comment is updated now.
To add a new dict of tensor, put it into the state dict directly with its name as key. To add a new tensor, put it into the sub-dict with key `"regular_tensors"`.

**Comment by frozenbugs** [comment]
@RamonZhou can you resolve the conflict and re-run the CI?

File: tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py [hunk 5]
```diff
@@ -798,6 +912,98 @@ def test_hetero_graph_on_shared_memory(
...
+
+
+@unittest.skipIf(
+    F._default_context_str == "gpu",
+    reason="Graph is CPU only at present.",
+)
+def test_multiprocessing_with_shared_memory():
+    """Test if two CSCSamplingGraphs are on the same shared memory after spawning.
+
+    For now this code only works when the sharing strategy of torch.multiprocessing is set to `file_system` at the beginning.
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:944]: line too long.

File: graphbolt/src/csc_sampling_graph.cc [hunk 6]
```diff
@@ -97,6 +97,48 @@ void CSCSamplingGraph::Save(torch::serialize::OutputArchive& archive) const {
   }
 }
 
+void CSCSamplingGraph::SetState(
+    torch::Dict<std::string, torch::Dict<std::string, torch::Tensor>>& state) {
+  auto&& regular_tensors = state.at("regular_tensors");
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:102]: worth to add a comment to explain why we use && here.

File: graphbolt/src/csc_sampling_graph.cc [hunk 7]
```diff
@@ -97,6 +97,48 @@ void CSCSamplingGraph::Save(torch::serialize::OutputArchive& archive) const {
...
+  TORCH_CHECK(
+      regular_tensors.at("magic_number").equal(torch::tensor({1})),
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:104]: Add `static const int kPickleVersion = 6199` at L24 then:
```suggestion
      regular_tensors.at("version").equal(torch::tensor({kPickleVersion})),
```

File: graphbolt/src/csc_sampling_graph.cc (hunk 6 above)
Comment (frozenbugs) [reply graphbolt/src/csc_sampling_graph.cc:102]: What about just 
```suggestion
  auto&& tensors = state.at("tensors");
```

File: graphbolt/src/csc_sampling_graph.cc (hunk 7 above)
Comment (frozenbugs) [reply graphbolt/src/csc_sampling_graph.cc:104]: Update GetState accordingly.

File: graphbolt/src/csc_sampling_graph.cc [hunk 8]
```diff
@@ -97,6 +97,48 @@ void CSCSamplingGraph::Save(torch::serialize::OutputArchive& archive) const {
...
+  }
+}
+
+torch::Dict<std::string, torch::Dict<std::string, torch::Tensor>>
+CSCSamplingGraph::GetState() const {
+  // State is a dict of dicts. The tensor-type attributes are stored in the dict
+  // with key "regular_tensors". The dict-type attributes (edge_attributes) are
+  // stored directly with the their name as the key.
+  torch::Dict<std::string, torch::Dict<std::string, torch::Tensor>> state;
+  torch::Dict<std::string, torch::Tensor> regular_tensors;
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:125]:
```suggestion
  torch::Dict<std::string, torch::Tensor> independent_tensors;
```

File: graphbolt/src/csc_sampling_graph.cc [hunk 9]
```diff
@@ -97,6 +97,48 @@ void CSCSamplingGraph::Save(torch::serialize::OutputArchive& archive) const {
...
+  regular_tensors.insert("magic_number", torch::tensor({1}));  // Magic number.
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:126]: Why we only need magic_number in regular_tensors?

File: tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py (hunk 3 above)
Comment (frozenbugs) [reply tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:258]: @RamonZhou is this addressed?

File: tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py [hunk 10]
```diff
@@ -251,6 +254,115 @@ def test_load_save_hetero_graph(num_nodes, num_edges, num_ntypes, num_etypes):
...
+        csc_indptr,
+        indices,
+        node_type_offset,
+        type_per_edge,
+        metadata,
+    ) = gbt.random_hetero_graph(num_nodes, num_edges, num_ntypes, num_etypes)
+    edge_attributes = {
+        "a": torch.randn((num_edges,)),
+    }
+    graph = gb.from_csc(
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:352]: ditto

[the remark this answers is not in the document]
Comment (RamonZhou) [reply tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:352]: Yes. I've added `test_multiprocessing_with_shared_memory`

File: tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py (hunk 3 above)
Comment (RamonZhou) [reply tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:258]: Yes. I've added `test_multiprocessing`.

File: graphbolt/src/csc_sampling_graph.cc (hunk 6 above)
Comment (RamonZhou) [reply graphbolt/src/csc_sampling_graph.cc:102]: I think `const auto &` is enough. No need to use `&&`.

File: graphbolt/src/csc_sampling_graph.cc (hunk 9 above)
Comment (RamonZhou) [reply graphbolt/src/csc_sampling_graph.cc:126]: The magic number (or version number) is for the whole state. Only is regular_tensors a good place to put it.

**Review by frozenbugs — APPROVED** [review]

File: tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py [hunk 11]
```diff
@@ -786,6 +900,100 @@ def test_hetero_graph_on_shared_memory(
...
+    graph_queue = ctx.Queue()  # Used for sending graph.
+    flag_queue = ctx.Queue()  # Used for sending finish signal.
+
+    p = ctx.Process(
+        target=process_csc_sampling_graph_with_queue,
+        args=(graph, graph_queue, flag_queue),
+    )
+    p.start()
+    try:
+        graph2 = graph_queue.get()  # Get the graph from the other process.
```
Comment (czkkkkkk) [inline tests/python/pytorch/graphbolt/impl/test_csc_sampling_graph.py:975]: If you get the graph from a subprocess through queue, the graph is on the same shared memory as your test graph. Then the testing is meaningless. I'd suggest you to randomly assign values to the tensors of the graph in the subprocess and test whether the test graph in the main process has the same values.

File: graphbolt/include/graphbolt/csc_sampling_graph.h [hunk 12]
```diff
@@ -129,6 +129,20 @@ class CSCSamplingGraph : public torch::CustomClassHolder {
    */
   void Save(torch::serialize::OutputArchive& archive) const;
 
+  /**
+   * @brief Pickle method for deserializing.
+   * @param state The state of serialized CSCSamplingGraph.
+   */
+  void SetState(
+      torch::Dict<std::string, torch::Dict<std::string, torch::Tensor>>& state);
```
[the remark this answers is not in the document]
Comment (RamonZhou) [reply graphbolt/include/graphbolt/csc_sampling_graph.h]: Added

**Review by czkkkkkk — APPROVED** [review]

**Pull request merged**